pkg/kubectl/cmd/util/factory.go

```diff
@@ -197,32 +199,57 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 		clients: clients,
 		flags:   flags,
 
-		Object: func() (meta.RESTMapper, runtime.ObjectTyper) {
+		// If discoverDynamicAPIs is true, make API calls to the discovery service to find APIs that
+		// have been dynamically added to the apiserver
+		Object: func(discoverDynamicAPIs bool) (meta.RESTMapper, runtime.ObjectTyper) {
 			cfg, err := clientConfig.ClientConfig()
 			CheckErr(err)
 			cmdApiVersion := unversioned.GroupVersion{}
 			if cfg.GroupVersion != nil {
 				cmdApiVersion = *cfg.GroupVersion
 			}
+			CheckErr(err)
+			if discoverDynamicAPIs {
+				client, err := clients.ClientForVersion(&unversioned.GroupVersion{Version: "v1"})
+				CheckErr(err)
+				
+				versions, gvks, err := GetThirdPartyGroupVersions(client.Discovery())
+				CheckErr(err)
+				if len(versions) > 0 {
+					preferredExternalVersion := versions[0]
```

Comment:
This block looks generally useful, can you factor it out into a method?  `BuildThirdThirdPartyRESTMapper(versions, gvks) RESTMapper`